File: go/internal/domain/time.go
package domain

import "time"

// EnsureUTC は時刻を UTC に正規化する。
//
// 内部で扱う時刻はすべて UTC に揃える。ローカル時刻のまま計算すると、夏時間の
// 切り替え日に請求期間が 23 時間や 25 時間になる。Go の time.Time は常に位置情報を
// 持つので、Python 版のような naive/aware の区別は不要だが、UTC への正規化は要る。
func EnsureUTC(t time.Time) time.Time {
	return t.UTC()
}

// AddMonths は月を加算する。
//
// 1/31 の 1 か月後は 2/28（閏年なら 2/29）とする。標準の AddDate は 2/31 を 3/3 に
// 繰り上げてしまい、課金サイクルとしては誤りになる。月末の扱いは仕様そのものなので、
// ライブラリ任せにせずここに書いて明示する。
func AddMonths(t time.Time, months int) time.Time {
	t = t.UTC()
	year, month, day := t.Date()
	total := int(month) - 1 + months
	year += total / 12
	newMonth := time.Month(total%12 + 1)
	if total%12 < 0 {
		year--
		newMonth = time.Month(total%12 + 13)
	}
	if last := daysIn(year, newMonth); day > last {
		day = last
	}
	return time.Date(year, newMonth, day, t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), time.UTC)
}

func daysIn(year int, month time.Month) int {
	return time.Date(year, month+1, 0, 0, 0, 0, 0, time.UTC).Day()
}

File: python/src/billing/domain/time.py
"""ドメインが扱う時刻の規約。

内部で扱う ``datetime`` は必ず tz-aware で、かつ UTC に正規化する。ローカル時刻の
まま計算すると、夏時間の切り替え日に請求期間が 23 時間や 25 時間になる。
"""

from __future__ import annotations

import calendar
from datetime import UTC, datetime

from billing.domain.errors import InvariantViolation


def ensure_aware(value: datetime, *, field: str) -> datetime:
    """tz-aware であることを確かめ、UTC に正規化して返す。"""
    if value.tzinfo is None or value.tzinfo.utcoffset(value) is None:
        raise InvariantViolation(f"{field} must be timezone-aware, got naive {value.isoformat()}")
    return value.astimezone(UTC)


def add_months(value: datetime, months: int) -> datetime:
    """月を加算する。

    1/31 の 1 か月後は 2/28（閏年なら 2/29）とする。日付をそのまま足すと存在しない
    日ができるため、月末に丸める。この「月末の扱い」は課金サイクルの仕様そのもので
    あり、ライブラリ任せにせずドメインに書いて明示する。
    """
    total = (value.year * 12 + (value.month - 1)) + months
    year, month = divmod(total, 12)
    month += 1
    day = min(value.day, calendar.monthrange(year, month)[1])
    return value.replace(year=year, month=month, day=day)
